share/availability/discovery/discovery.go

```diff
@@ -183,23 +199,42 @@ func (d *Discovery) ensurePeers(ctx context.Context) {
 			log.Info("Context canceled. Finishing peer discovery")
 			return
 		case <-t.C:
-			if uint(d.set.Size()) == d.peersLimit {
-				// stop ticker if we have reached the limit
-				t.Stop()
-				continue
-			}
-			peers, err := d.disc.FindPeers(ctx, topic)
-			if err != nil {
-				log.Error(err)
-				continue
-			}
-			for p := range peers {
-				go d.handlePeerFound(ctx, topic, p)
-			}
+			d.findPeers(ctx)
 		}
 	}
 }
 
+func (d *Discovery) findPeers(ctx context.Context) {
+	if uint(d.set.Size()) >= d.peersLimit {
+		log.Debugw("at peer limit, skipping FindPeers", "size", d.set.Size())
+		return
+	}
+
+	peers, err := d.disc.FindPeers(ctx, topic)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	// we use a wait group to avoid a context leak
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(ctx)
+
+	for p := range peers {
+		wg.Add(1)
+		go func(peer peer.AddrInfo) {
+			defer wg.Done()
+			d.handlePeerFound(ctx, peer, cancel)
+		}(p)
+	}
+
+	// Wait for all goroutines to finish, then cancel the context
+	go func() {
+		wg.Wait()
+		cancel()
+	}()
+}
```

Comment:
The peers channel is returned immediately and closed when either query ends, or the context to FindPeers is canceled. The query can take minutes, especially when the network is significant, and many peers are undialable( e.g., in DHT server mode but behind a NAT). 

So if we obtain all the necessary connections, let's cancel FindPeers as we cancel all outgoing connections once we no longer need them. In other words, current logic works well when FindPeers ends earlier than connections are obtained, but not vice-versa.